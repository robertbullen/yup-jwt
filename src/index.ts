import * as jwt from 'jsonwebtoken';
import * as yup from 'yup';

// Declare a `jwt()` method on the `StringSchema` interface via module augmentation.
declare module 'yup' {
    export interface StringSchema {
        jwt(message?: string): StringSchema;
    }
}

// tslint:disable-next-line:no-invalid-template-strings
export const defaultMessage: string = '${path} must be a parsable JSON web token';

// Implement the `StringSchema.jwt()` method.
yup.addMethod(yup.string, 'jwt', function(this: yup.StringSchema, message: string = defaultMessage): yup.StringSchema {
    return this.test({
        exclusive: true,
        message,
        name: 'jwt',
        test(this: yup.TestContext, value: string): boolean {
            return !!jwt.decode(value);
        },
    });
});
