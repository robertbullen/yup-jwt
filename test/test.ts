import 'jest';
import * as jwt from 'jsonwebtoken';
import * as yup from 'yup';

import { defaultMessage } from '../src/index';

describe('`yup.string().jwt()`', () => {
    const nonsenseToken: string = 'foo.bar.qux';
    const schema: yup.StringSchema = yup.string().jwt();

    describe('`isValidSync()`', () => {
        const stringPayload: string = 'payload';
        const objectPayload: object = { stringPayload };
        const secret: string = 'secret';

        describe('returns true', () => {
            it('when testing a valid token with a string payload', () => {
                const token: string = jwt.sign(stringPayload, secret);
                expect(schema.isValidSync(token)).toBe(true);
            });

            it('when testing a valid token with an object payload', () => {
                const token: string = jwt.sign(objectPayload, secret);
                expect(schema.isValidSync(token)).toBe(true);
            });

            it('when testing a valid token that has expired', () => {
                const payload = { stringPayload, exp: Math.floor(Date.now() / 1000) - 30 };
                const token: string = jwt.sign(payload, secret);
                expect(schema.isValidSync(token)).toBe(true);
            });
        });

        describe('returns false', () => {
            it('when testing an nonsense token', () => {
                expect(schema.isValidSync(nonsenseToken)).toBe(false);
            });

            it('when testing a corrupted token', () => {
                const token: string = jwt.sign(stringPayload, secret);
                const corruptedToken: string = token + '#';
                expect(schema.isValidSync(corruptedToken)).toBe(false);
            });
        });
    });

    describe('`validateSync()`', () => {
        describe('throws an error', () => {
            it('with the default message when no custom message is supplied', () => {
                // tslint:disable-next-line:no-invalid-template-strings
                const expectedMessage: string = defaultMessage.replace('${path}', 'this');
                expect(() => schema.validateSync(nonsenseToken)).toThrowError(expectedMessage);
            });

            it('with a custom message when one is supplied', () => {
                const customMessage: string = 'custom message';
                const customMessageSchema: yup.StringSchema = yup.string().jwt(customMessage);
                expect(() => customMessageSchema.validateSync(nonsenseToken)).toThrowError(customMessage);
            });
        });
    });
});
