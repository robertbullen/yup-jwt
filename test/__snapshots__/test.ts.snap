// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yup.string().jwt('decode') validateSync() throws an error with the default message when no custom message is supplied 1`] = `"this must be a decodable JSON web token"`;

exports[`yup.string().jwt('regex') validateSync() throws an error with the default message when no custom message is supplied 1`] = `"this must be a JSON web token that satisfies the pattern /^[0-9a-z\\\\-_]+?\\\\.[0-9a-z\\\\-_]+?\\\\.([0-9a-z\\\\-_]+)?$/i"`;
